src/exec.js

```diff
@@ -37,47 +37,29 @@ function execSync(cmd, opts, pipe) {
     encoding: 'utf8',
   }, opts);
 
-  if (fs.existsSync(scriptFile)) common.unlinkSync(scriptFile);
-  if (fs.existsSync(stdoutFile)) common.unlinkSync(stdoutFile);
-  if (fs.existsSync(stderrFile)) common.unlinkSync(stderrFile);
   if (fs.existsSync(codeFile)) common.unlinkSync(codeFile);
-
-  var execCommand = JSON.stringify(common.config.execPath) + ' ' + JSON.stringify(scriptFile);
-  var script;
+  if (fs.existsSync(paramsFile)) common.unlinkSync(paramsFile);
+  if (fs.existsSync(stderrFile)) common.unlinkSync(stderrFile);
+  if (fs.existsSync(stdoutFile)) common.unlinkSync(stdoutFile);
 
   opts.cwd = path.resolve(opts.cwd);
-  var optString = JSON.stringify(opts);
-
-  script = [
-    "var child = require('child_process')",
-    "  , fs = require('fs');",
-    'var childProcess = child.exec(' + JSON.stringify(cmd) + ', ' + optString + ', function(err) {',
-    '  var fname = ' + JSON.stringify(codeFile) + ';',
-    '  if (!err) {',
-    '    fs.writeFileSync(fname, "0");',
-    '  } else if (err.code === undefined) {',
-    '    fs.writeFileSync(fname, "1");',
-    '  } else {',
-    '    fs.writeFileSync(fname, err.code.toString());',
-    '  }',
-    '});',
-    'var stdoutStream = fs.createWriteStream(' + JSON.stringify(stdoutFile) + ');',
-    'var stderrStream = fs.createWriteStream(' + JSON.stringify(stderrFile) + ');',
-    'childProcess.stdout.pipe(stdoutStream, {end: false});',
-    'childProcess.stderr.pipe(stderrStream, {end: false});',
-    'childProcess.stdout.pipe(process.stdout);',
-    'childProcess.stderr.pipe(process.stderr);',
-  ].join('\n') +
-    (pipe ? '\nchildProcess.stdin.end(' + JSON.stringify(pipe) + ');\n' : '\n') +
-    [
-      'var stdoutEnded = false, stderrEnded = false;',
-      'function tryClosingStdout(){ if(stdoutEnded){ stdoutStream.end(); } }',
-      'function tryClosingStderr(){ if(stderrEnded){ stderrStream.end(); } }',
-      "childProcess.stdout.on('end', function(){ stdoutEnded = true; tryClosingStdout(); });",
-      "childProcess.stderr.on('end', function(){ stderrEnded = true; tryClosingStderr(); });",
-    ].join('\n');
-
-  fs.writeFileSync(scriptFile, script);
+
+  var paramsToSerialize = {
+    command: cmd,
+    execOptions: opts,
+    pipe: pipe,
+    stdoutFile: stdoutFile,
+    stderrFile: stderrFile,
+    codeFile: codeFile,
+  };
+
+  fs.writeFileSync(paramsFile, JSON.stringify(paramsToSerialize), 'utf8');
```

Comment:
What is the advantage of storing the params in a separate file, as opposed to simply passing them in as a argument to the command?

Reply:
That was my original approach, but there's some issues with passing via CLI argument (at least in shells, via `execSync`). As an example, we could look at the multi-line string:

```
has"chars
.txt
```

`JSON.stringify` gives us `"has\"chars\n.txt"`. So far, so good.

When this gets passed to the shell, it looks something like:

```sh
node exec-child.js "has\"chars\n.txt"
```

The shell interprets the string and supplies the interpreted value to the node process, so `process.argv[2]` looks like `has"chars\n.txt` (that's a one-line string with a literal `\` followed by a literal `n`).

This is where we hit issues. Our string is partially deserialized, so we can't use `JSON.parse`, but it's still not equivalent to our original string. This is an unfortunate consequence of using both shell and JS--they have different opinions on how to escape strings.

---

We can consider passing these by CLI argument if we replace `execSync()` with `execFileSync()` within `exec.js`. This is one of my intended future optimizations, but I left it out to limit complexity. Such a change is backwards compatible (we just can't change `exec()` within `exec-child.js`).

Reply:
Ok, that makes sense. We eliminate the need to write the JS file every call, but in return we need to write the arguments to a file. Not the best, but it's an improvement overall.

Reply:
This is still strictly better. The JS file used to contain each argument written inside it (as literals), and `sizeof(code) + sizeof(arguments) > sizeof(arguments)`.

I'll also send some PRs to trim down the number of arguments, which will be a bigger win.